### DISCUSSION TITLE
Replace `assert` with `if` condition in `checkAndCalculateBrokerFee`

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/evm-monorepo#854] [General]
There is only one place where we still use `assert` check. 
https://github.com/sablier-labs/v2-core/blob/be1dea483581eefa4a048369853db15404fe1e07/src/libraries/Helpers.sol#L42

Integration tests also don't cover this scenario where broker fee exceeds total amount. This seems to be safe because we already have a check above it 
https://github.com/sablier-labs/v2-core/blob/be1dea483581eefa4a048369853db15404fe1e07/src/libraries/Helpers.sol#L33-L35

However, since we are re-checking whether broker amount is less than total amount, we should also cover this in tests and replae it it with `if`.

```solidity
  // Assert that the the broker fee amount is strictly less than the total amount.
  if (amounts.brokerFee >= totalAmount) {
      revert Errors.SablierV2Lockup_BrokerFeeExceedsTotalAmount(brokerFee, totalAmount);
  }
```

This will also increase branch coverage of Helpers.sol.

<img width="901" alt="Screenshot 2024-03-17 at 16 29 06" src="https://github.com/sablier-labs/v2-core/assets/6676622/01b18a7a-05a0-4597-97f8-cd4f6151c791">

Would love to hear your thoughts on this @PaulRBerg.

**Comment by PaulRBerg** [comment]
Since the max fee is [hard-coded to 10%](https://github.com/sablier-labs/v2-core/blob/be1dea483581eefa4a048369853db15404fe1e07/src/abstracts/SablierV2Lockup.sol#L36), it is impossible to write a test that makes the `amounts.brokerFee` greater than `totalAmount`. That is, unless we deploy a mock contract with a different max fee (or change the value with `StdStorage`?), but there's no point in doing this because the max fee is hard-coded.

This assertion is there on a precautionary basis - if there's an unknown bug, the assertion will stop contract execution. But since we don't know what bug there is, we cannot write a test for it.

Given this rationale, I'd keep the assertion as is.

P.S. See the discussion I opened on Twitter about production assertions: [What are your thoughts on leaving asserts in production contracts?](https://twitter.com/PaulRBerg/status/1618608953352675328)

**Reply by smol-ninja** [reply]
Fair. I understand your intention here. Having said that, it still looks superfluous to re-validate it right after an arithmetic operation. The two validations inside `checkAndCalculateBrokerFee` (`brokerFee.gt(maxBrokerFee)` and `assert(totalAmount > amounts.brokerFee)`) only have a multiplication operation between them. So, this validates whether `.mul`, `ud` and `intoUint256` are working as expected or not.

What if we move this `assert` in `_createWithTimestamps` function after the following line?

```solidity
Helpers.checkCreateWithTimestamps(createAmounts.deposit, params.segments, MAX_SEGMENT_COUNT, params.startTime);
```
Then the assertion is made in the context of the create function and not the helper function.

PS: I am only brainstorming to understand if the invariant is checked at the right place. I have already agreed with having assertions here.

**Comment by PaulRBerg** [comment]
On a separate note, this discussion makes me realize that it would be helpful to write a test that validates that `MAX_BROKER_FEE` has a value of 10%. WDYT?

**Reply by smol-ninja** [reply]
Though an interesting take on validating the constant values, I think it would be unnecessary. 10% is just a value assigned to a constant variable `MAX_BROKER_FEE` (unless we assume that constant can be changed too). Why made you realize this?

**Reply by PaulRBerg** [reply]
We might accidentally edit the constant value, and it isn't something that is obviously noticeable. Even an auditor might assume that the change was intentional.

With a test, we would detect the accidental edit.

**Reply by smol-ninja** [reply]
Fair enough.

**Comment by smol-ninja** [comment]
Closing it since no more discussion is required.
